app/controllers/invites_controller.rb

```diff
@@ -109,73 +78,104 @@ def create
+    Invite.transaction do
+      invite.topic_invites.destroy_all
+      invite.topic_invites.create!(topic_id: topic.id) if topic.present?
 
-        if topic.present?
-          guardian.ensure_can_invite_to!(topic)
-        else
-          raise Discourse::InvalidParameters.new(:topic_id)
-        end
-      end
-    end
+      invite.invited_groups.destroy_all
+      groups.each { |group| invite.invited_groups.find_or_create_by!(group_id: group.id) } if groups.present?
 
-    invite_link = if is_single_invite
-      Invite.generate_single_use_invite_link(params[:email], current_user, topic, group_ids)
-    else
-      Invite.generate_multiple_use_invite_link(
-        invited_by: current_user,
+      invite.update!(
         max_redemptions_allowed: params[:max_redemptions_allowed],
-        expires_at: params[:expires_at],
-        group_ids: group_ids
+        expires_at: params[:expires_at]
       )
     end
-    if invite_link.present?
-      render_json_dump(invite_link)
-    else
-      render json: failed_json, status: 422
-    end
-  rescue => e
-    render json: { errors: [e.message] }, status: 422
+
+    render json: success_json
   end
 
   def destroy
     params.require(:id)
 
     invite = Invite.find_by(invited_by_id: current_user.id, id: params[:id])
     raise Discourse::InvalidParameters.new(:id) if invite.blank?
+
     invite.trash!(current_user)
 
     render json: success_json
   end
 
+  def perform_accept_invitation
+    params.require(:id)
+    params.permit(:email, :username, :name, :password, :timezone, user_custom_fields: {})
+
+    invite = Invite.find_by(invite_key: params[:id])
+
+    if invite.present?
+      begin
+        user = invite.redeem(
+          email: invite.is_invite_link? ? params[:email] : invite.email,
+          username: params[:username],
+          name: params[:name],
+          password: params[:password],
+          user_custom_fields: params[:user_custom_fields],
+          ip_address: request.remote_ip
+        )
+
+        if user.present?
+          log_on_user(user) if user.active?
+          user.update_timezone_if_missing(params[:timezone])
+          post_process_invite(user)
+          response = { success: true }
+        else
+          response = { success: false, message: I18n.t('invite.not_found_json') }
+        end
+
+        topic = invite.topics.first
+        if user.present? && user.active?
+          response[:redirect_to] = topic.present? ? path(topic.relative_url) : path("/")
+        elsif user.present?
+          response[:message] = I18n.t('invite.confirm_email')
+          cookies[:destination_url] = path(topic.relative_url) if topic.present?
+        end
+
+        render json: response
+      rescue ActiveRecord::RecordInvalid, ActiveRecord::RecordNotSaved => e
+        render json: {
+          success: false,
+          errors: e.record&.errors&.to_hash || {},
+          message: I18n.t('invite.error_message')
+        }
+      rescue Invite::UserExists => e
```

Comment:
I think the errors that we're trying to rescue only affects the `Invite#redeem` method call. We can just wrap that method call in the `begin..rescue..end` block and not the rest of the code.